prometheus-proc/src/Prometheus/Metric/Proc.hs

```diff
@@ -0,0 +1,175 @@
+{-# language ForeignFunctionInterface #-}
+{-# language OverloadedStrings #-}
+{-# language RecordWildCards #-}
+
+module Prometheus.Metric.Proc ( procMetrics ) where
+
+import Data.Char ( isSpace )
+import Data.Maybe ( catMaybes )
+import Data.String ( fromString )
+import Foreign.C
+import Prometheus
+import System.FilePath
+import System.IO.Unsafe
+import System.Posix.Memory ( sysconfPageSize )
+import System.Posix.Process ( getProcessID )
+import qualified Text.Regex.Applicative as RE
+import qualified Text.Regex.Applicative.Common as RE
+
+
+data ProcMetrics =
+  ProcMetrics
+
+
+procMetrics :: Prometheus.Metric ProcMetrics
+procMetrics =
+  Metric ( return ( ProcMetrics, collect ) )
+
+
+foreign import ccall unsafe
+  clk_tck :: CLong
+
+
+collect :: IO [ SampleGroup ]
+collect = do
+  pid <-
+    getProcessID
+
+  mprocStat <-
+      RE.match parseProcStat <$> readFile ( "/" </> "proc" </> show pid </> "stat" )
+
+  return
+    ( foldMap ( toMetrics ) mprocStat )
+
+
+toMetrics :: ProcStat -> [ SampleGroup ]
+toMetrics ProcStat{..} =
+  catMaybes
+    [ Just process_cpu_seconds_total
+    , process_start_time_seconds
+    , Just process_virtual_memory_bytes
+    , Just process_resident_memory_bytes
+    ]
+
+  where
+
+    process_cpu_seconds_total =
+      metric
+        "process_cpu_seconds_total"
+        "Total user and system CPU time spent in seconds."
+        CounterType
+        ( fromTicks ( utime + stime ) )
+
+    process_start_time_seconds = do
+      btime <-
+        mbtime
+
+      return
+        ( metric
+            "process_start_time_seconds"
+            "Start time of the process since unix epoch in seconds."
+            GaugeType
+            ( fromIntegral btime + fromTicks starttime )
+        )
+
+    process_virtual_memory_bytes =
+      metric
+        "process_virtual_memory_bytes"
+        "Virtual memory size in bytes."
+        GaugeType
+        vsize
+
+    process_resident_memory_bytes =
+      metric
+        "process_resident_memory_bytes"
+        "Resident memory size in bytes."
+        GaugeType
+        ( rss * sysconfPageSize )
+
+    fromTicks ticks =
+      fromIntegral ticks / fromIntegral clk_tck :: Double
```

Comment:
This function needs a comment, I think. I don't know what ticks are, and can't readily tell what they are being converted to.

Reply:
At first my response would be "well the contents of this file work with `/proc` so you're expected to just know it", but there's no reason it should be that way! I'll promote this to the top-level and document what it's doing and what a tick is (I very much agree with you).